controllers/route/route_certificaterequest_controller.go

```diff
@@ -0,0 +1,362 @@
+package route
+
+import (
+	"bytes"
+	"context"
+	"crypto"
+	"crypto/rand"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/pem"
+	"sort"
+	"time"
+
+	"github.com/go-logr/logr"
+	cmutil "github.com/jetstack/cert-manager/pkg/api/util"
+	cmapi "github.com/jetstack/cert-manager/pkg/apis/certmanager/v1"
+	cmmeta "github.com/jetstack/cert-manager/pkg/apis/meta/v1"
+	cmutilpki "github.com/jetstack/cert-manager/pkg/util/pki"
+	routev1 "github.com/openshift/api/route/v1"
+	outils "github.com/redhat-cop/operator-utils/pkg/util"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+type CertificateRequestReconciler struct {
+	outils.ReconcilerBase
+	Log            logr.Logger
+	controllerName string
+}
+
+func (c *CertificateRequestReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	c.controllerName = "route_certificaterequest_controller"
+
+	isCertManagerRoute := predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			route, ok := e.Object.DeepCopyObject().(*routev1.Route)
+			if !ok {
+				return false
+			}
+			return hasCertManagerAnnotations(route)
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			route, ok := e.Object.DeepCopyObject().(*routev1.Route)
+			if !ok {
+				return false
+			}
+			return hasCertManagerAnnotations(route)
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			newRoute, ok := e.ObjectNew.DeepCopyObject().(*routev1.Route)
+			if !ok {
+				return false
+			}
+			return hasCertManagerAnnotations(newRoute)
+		},
+		GenericFunc: func(e event.GenericEvent) bool {
+			return false
+		},
+	}
+
+	hasCRStatusChanged := predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			_, ok := e.Object.(*cmapi.CertificateRequest)
+			return ok
```

Comment:
This reams superfluous, it is only ever a  CertificateRequest

Reply:
I agree, but this ultra-defensive style is in the other controller in the original repo, so I left it as so.